openmc/filter.py

```diff
@@ -0,0 +1,1810 @@
+
+
+class MeshSurfaceFilter(MeshFilter):
+    """Filter events by surface crossings on a regular, rectangular mesh.
+
+    Parameters
+    ----------
+    mesh : openmc.MeshBase
+        The mesh object that events will be tallied onto
+    filter_id : int
+        Unique identifier for the filter
+
+    Attributes
+    ----------
+    bins : Integral
+        The mesh ID
+    mesh : openmc.MeshBase
+        The mesh object that events will be tallied onto
+    id : int
+        Unique identifier for the filter
+    bins : list of tuple
+
+        A list of mesh indices / surfaces for each filter bin, e.g. [(1, 1,
+        'x-min out'), (1, 1, 'x-min in'), ...]
+
+    num_bins : Integral
+        The number of filter bins
+
+    """
+
+    @MeshFilter.mesh.setter
+    def mesh(self, mesh):
+        cv.check_type('filter mesh', mesh, openmc.MeshBase)
+        self._mesh = mesh
+
+        # Take the product of mesh indices and current names
+        n_dim = mesh.n_dimension
+        self.bins = [mesh_tuple + (surf,) for mesh_tuple, surf in
+                     product(mesh.indices, _CURRENT_NAMES[:4*n_dim])]
+
+    def get_pandas_dataframe(self, data_size, stride, **kwargs):
+        """Builds a Pandas DataFrame for the Filter's bins.
+
+        This method constructs a Pandas DataFrame object for the filter with
+        columns annotated by filter bin information. This is a helper method for
+        :meth:`Tally.get_pandas_dataframe`.
+
+        Parameters
+        ----------
+        data_size : int
+            The total number of bins in the tally corresponding to this filter
+        stride : int
+            Stride in memory for the filter
+
+        Returns
+        -------
+        pandas.DataFrame
+            A Pandas DataFrame with three columns describing the x,y,z mesh
+            cell indices corresponding to each filter bin.  The number of rows
+            in the DataFrame is the same as the total number of bins in the
+            corresponding tally, with the filter bin appropriately tiled to map
+            to the corresponding tally bins.
+
+        See also
+        --------
+        Tally.get_pandas_dataframe(), CrossFilter.get_pandas_dataframe()
+
+        """
+        # Initialize Pandas DataFrame
+        df = pd.DataFrame()
+
+        # Initialize dictionary to build Pandas Multi-index column
+        filter_dict = {}
+
+        # Append mesh ID as outermost index of multi-index
+        mesh_key = 'mesh {}'.format(self.mesh.id)
+
+        # Find mesh dimensions - use 3D indices for simplicity
+        n_surfs = 4 * len(self.mesh.dimension)
+        if len(self.mesh.dimension) == 3:
+            nx, ny, nz = self.mesh.dimension
+        elif len(self.mesh.dimension) == 2:
+            nx, ny = self.mesh.dimension
+            nz = 1
+        else:
+            nx = self.mesh.dimension
+            ny = nz = 1
+
+        # Generate multi-index sub-column for x-axis
+        filter_bins = np.arange(1, nx + 1)
```

Comment:
I see that these 6 lines of code are repeated here and in multiple places. I would try and modularize it into a function that takes in the `repeat_factor` and the `key` (x, y, or z) as well as the `filter_dict` object. It could fill in values for `filter_dict` as per the logic specified (as it only seems repeat_factor changes between all the useages of these lines of code). 

This might cut down a significant amount of repetition in this file.